File: desktop/src-tauri/src/managed_agents/discovery/presets.rs
use std::path::PathBuf;

use crate::managed_agents::{
    AcpAvailabilityStatus, AcpRuntimeCatalogEntry, AuthStatus, HarnessSource,
};

use super::normalize_agent_args;

pub(super) struct PresetHarness {
    pub(super) id: &'static str,
    pub(super) label: &'static str,
    pub(super) command: &'static str,
    pub(super) args: &'static [&'static str],
    pub(super) install_instructions_url: &'static str,
    pub(super) install_hint: &'static str,
    pub(super) underlying_cli: Option<&'static str>,
}

pub(super) fn preset_catalog_entry(
    def: &PresetHarness,
    resolve: impl Fn(&str) -> Option<PathBuf>,
) -> AcpRuntimeCatalogEntry {
    let (availability, command, binary_path) = match resolve(def.command) {
        Some(path) => (
            AcpAvailabilityStatus::Available,
            Some(def.command.to_string()),
            Some(path.display().to_string()),
        ),
        None => {
            let underlying_cli_found = def
                .underlying_cli
                .map(|cli| resolve(cli).is_some())
                .unwrap_or(false);
            if underlying_cli_found {
                (AcpAvailabilityStatus::AdapterMissing, None, None)
            } else {
                (AcpAvailabilityStatus::NotInstalled, None, None)
            }
        }
    };
    let underlying_cli_path = def
        .underlying_cli
        .and_then(resolve)
        .map(|path| path.display().to_string());
    let default_args = normalize_agent_args(
        def.command,
        def.args.iter().map(|arg| arg.to_string()).collect(),
    );

    AcpRuntimeCatalogEntry {
        id: def.id.to_string(),
        label: def.label.to_string(),
        avatar_url: String::new(),
        availability,
        command,
        binary_path,
        default_args,
        mcp_command: None,
        model_env_var: None,
        provider_env_var: None,
        thinking_env_var: None,
        install_hint: def.install_hint.to_string(),
        install_instructions_url: def.install_instructions_url.to_string(),
        can_auto_install: false,
        requires_external_cli: false,
        underlying_cli_path,
        node_required: false,
        auth_status: AuthStatus::NotApplicable,
        login_hint: None,
        source: HarnessSource::Preset,
        definition_env: Default::default(),
    }
}

pub(super) const PRESET_HARNESSES: &[PresetHarness] = &[
    PresetHarness {
        id: "cursor",
        label: "Cursor",
        command: "cursor-agent",
        args: &["acp"],
        install_instructions_url: "https://cursor.com/downloads",
        install_hint: "Buzz talks to Cursor through the cursor-agent CLI's ACP mode.",
        underlying_cli: None,
    },
    PresetHarness {
        id: "omp",
        label: "Oh My Pi",
        command: "omp",
        args: &["acp"],
        install_instructions_url: "https://github.com/can1357/oh-my-pi",
        install_hint: "Buzz talks to Oh My Pi through its CLI's ACP mode (omp acp).",
        underlying_cli: None,
    },
    PresetHarness {
        id: "grok",
        label: "Grok Build",
        command: "grok",
        args: &["agent", "--always-approve", "stdio"],
        install_instructions_url: "https://build.x.ai/docs",
        install_hint: "Buzz talks to Grok Build through its CLI's agent stdio mode.",
        underlying_cli: None,
    },
    PresetHarness {
        id: "opencode",
        label: "OpenCode",
        command: "opencode",
        args: &["acp"],
        install_instructions_url: "https://opencode.ai/docs",
        install_hint: "Buzz talks to OpenCode through its CLI's ACP mode (opencode acp).",
        underlying_cli: None,
    },
    PresetHarness {
        id: "kimi",
        label: "Kimi Code",
        command: "kimi",
        args: &["acp"],
        install_instructions_url: "https://kimi.ai/download",
        install_hint: "Buzz talks to Kimi Code through its CLI's ACP mode (kimi acp).",
        underlying_cli: None,
    },
    PresetHarness {
        id: "amp",
        label: "Amp",
        command: "amp-acp",
        args: &[],
        install_instructions_url: "https://github.com/tao12345666333/amp-acp",
        install_hint: "Buzz talks to the Amp CLI through the amp-acp adapter. Follow the setup guide to install the adapter so the amp-acp command is on your PATH.",
        underlying_cli: Some("amp"),
    },
    PresetHarness {
        id: "hermes",
        label: "Hermes Agent",
        command: "hermes-acp",
        args: &[],
        install_instructions_url: "https://hermes-agent.nousresearch.com",
        install_hint: "Buzz talks to Hermes Agent through its hermes-acp command.",
        underlying_cli: None,
    },
    PresetHarness {
        id: "openclaw",
        label: "OpenClaw",
        command: "openclaw",
        args: &["acp"],
        install_instructions_url: "https://docs.openclaw.ai/start/getting-started",
        install_hint: "Buzz talks to OpenClaw through its ACP mode (openclaw acp), which relies on the OpenClaw Gateway daemon. Follow the setup guide to install both.\n\n\
            ⚠️  Execution-locus note: `openclaw acp` runs tools inside the \
            OpenClaw Gateway daemon, not in the Desktop process. \
            Desktop-injected BUZZ_* env vars are visible to the `openclaw` \
            harness process itself, but do NOT automatically reach the \
            Gateway's execution environment. If your tools or agent logic \
            needs BUZZ_* credentials at execution time, set them on the \
            Gateway's own environment separately.",
        underlying_cli: None,
    },
    PresetHarness {
        id: "cybara",
        label: "Cybara",
        command: "cybara",
        args: &["acp"],
        install_instructions_url: "https://cybara.ai/download#cli-tui",
        install_hint: "Buzz talks to Cybara through its CLI's ACP mode (cybara acp).",
        underlying_cli: None,
    },
];

pub(crate) fn preset_harness_definitions(
) -> Vec<crate::managed_agents::custom_harnesses::HarnessDefinition> {
    PRESET_HARNESSES
        .iter()
        .map(
            |preset| crate::managed_agents::custom_harnesses::HarnessDefinition {
                id: preset.id.to_string(),
                label: preset.label.to_string(),
                command: preset.command.to_string(),
                args: preset.args.iter().map(|arg| arg.to_string()).collect(),
                env: std::collections::BTreeMap::new(),
                install_instructions_url: preset.install_instructions_url.to_string(),
                install_hint: preset.install_hint.to_string(),
            },
        )
        .collect()
}

pub(crate) fn preset_harness_ids() -> &'static [&'static str] {
    use std::sync::OnceLock;
    static IDS: OnceLock<Vec<&'static str>> = OnceLock::new();
    IDS.get_or_init(|| PRESET_HARNESSES.iter().map(|preset| preset.id).collect())
        .as_slice()
}

#[cfg(test)]
mod tests {
    use super::preset_harness_definitions;

    #[test]
    fn cybara_launches_native_acp_mode() {
        let cybara = preset_harness_definitions()
            .into_iter()
            .find(|preset| preset.id == "cybara")
            .expect("Cybara preset should exist");

        assert_eq!(cybara.label, "Cybara");
        assert_eq!(cybara.command, "cybara");
        assert_eq!(cybara.args, vec!["acp"]);
        assert_eq!(
            cybara.install_instructions_url,
            "https://cybara.ai/download#cli-tui"
        );
        assert!(cybara.install_hint.contains("cybara acp"));
        assert!(cybara.env.is_empty());
    }
}

File: beehive/src/presets.ts
import { accessSync, constants, realpathSync, statSync } from 'node:fs';
import { isAbsolute, join } from 'node:path';
/** Pinned Desktop definitions, not vendor capability or authentication claims.
 * 051c3a2 desktop/src-tauri/src/managed_agents/discovery/presets.rs:69–228.
 */
export const presets = [
  {
    "id": "pi",
    "label": "Pi",
    "command": "pi-acp",
    "args": [],
    "env": {},
    "installInstructionsUrl": "https://github.com/svkozak/pi-acp",
    "installHint": "Install Pi with npm install -g --ignore-scripts @earendil-works/pi-coding-agent. Install the Pi ACP adapter with npm install -g pi-acp.",
    "underlyingCli": "pi"
  },
  {
    "id": "devin",
    "label": "Devin",
    "command": "devin",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://docs.devin.ai/cli",
    "installHint": "Buzz talks to Devin through the official Devin CLI's ACP mode (devin acp)."
  },
  {
    "id": "cursor",
    "label": "Cursor",
    "command": "cursor-agent",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://cursor.com/downloads",
    "installHint": "Buzz talks to Cursor through the cursor-agent CLI's ACP mode."
  },
  {
    "id": "omp",
    "label": "Oh My Pi",
    "command": "omp",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://omp.sh/",
    "installHint": "Buzz talks to Oh My Pi through its CLI's ACP mode (omp acp)."
  },
  {
    "id": "grok",
    "label": "Grok Build",
    "command": "grok",
    "args": [
      "agent",
      "--always-approve",
      "stdio"
    ],
    "env": {},
    "installInstructionsUrl": "https://build.x.ai/docs",
    "installHint": "Buzz talks to Grok Build through its CLI's agent stdio mode."
  },
  {
    "id": "opencode",
    "label": "OpenCode",
    "command": "opencode",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://opencode.ai/docs",
    "installHint": "Buzz talks to OpenCode through its CLI's ACP mode (opencode acp)."
  },
  {
    "id": "kimi",
    "label": "Kimi Code",
    "command": "kimi",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://kimi.ai/download",
    "installHint": "Buzz talks to Kimi Code through its CLI's ACP mode (kimi acp)."
  },
  {
    "id": "amp",
    "label": "Amp",
    "command": "amp-acp",
    "args": [],
    "env": {},
    "installInstructionsUrl": "https://github.com/tao12345666333/amp-acp",
    "installHint": "Buzz talks to the Amp CLI through the amp-acp adapter. Follow the setup guide to install the adapter so the amp-acp command is on your PATH.",
    "underlyingCli": "amp"
  },
  {
    "id": "hermes",
    "label": "Hermes Agent",
    "command": "hermes-acp",
    "args": [],
    "env": {},
    "installInstructionsUrl": "https://hermes-agent.nousresearch.com",
    "installHint": "Buzz talks to Hermes Agent through its hermes-acp command."
  },
  {
    "id": "openclaw",
    "label": "OpenClaw",
    "command": "openclaw",
    "args": [
      "acp"
    ],
    "env": {},
    "installInstructionsUrl": "https://docs.openclaw.ai/start/getting-started",
    "installHint": "Install the amp-acp npm adapter.",
    "underlyingCli": "amp"
  }
] as const;
/** Process-free discovery in explicit local PATH only; never a login shell. */
export function discoverPresets(path: string = process.env.PATH ?? '') {
  const directories = path.split(':').filter(isAbsolute).slice(0, 128);
  const resolve = (command: string): string | undefined => {
    for (const directory of directories) {
      const candidate = join(directory, command);
      try { if (!statSync(candidate).isFile()) continue; accessSync(candidate, constants.X_OK); return realpathSync(candidate); }
      catch { /* Missing/inaccessible candidate is diagnostic absence, not auth. */ }
    }
    return undefined;
  };
  return presets.map(preset => {
    const executable = resolve(preset.command);
    const cli = 'underlyingCli' in preset ? resolve(preset.underlyingCli) : undefined;
    const availability = 'underlyingCli' in preset ? executable ? cli ? 'available' : 'cli-missing' : cli ? 'adapter-missing' : 'not-installed' : executable ? 'available' : 'executable-missing';
    return { ...preset, executable, availability, authentication: 'unverified' as const,
      reason: 'Setup/diagnostic only: pinned preset source provides no exact actual-session model and native profile contract. NotApplicable auth is not authenticated.' };
  });
}
/** Local display only: no execution, install scripts or remote credentials. */
export function showPresets(): void {
  for (const p of discoverPresets()) console.log(`${p.id} | ${p.label} | ${p.availability} | ${p.command} ${JSON.stringify(p.args)} | env {}\n${p.installHint} ${p.installInstructionsUrl}\n${p.reason}`);
  console.log('Sign in/configure locally as the dedicated host service user, not Desktop HOME. No login or ACP process was launched.');
}
